cgsmiles/resolve.py

```diff
@@ -11,38 +11,53 @@
 from .pysmiles_utils import (rebuild_h_atoms,
                              annotate_ez_isomers_cgsmiles)
 
-def compatible(left, right, legacy=True):
+def compatible(left, right, left_data={}, right_data={}, legacy=True):
     """
     Check bonding descriptor compatibility according
-    to the CGsmiles syntax conventions. With legacy
+    to the CGsmiles syntax conventions. With legacy,
     the BigSmiles convention can be used.
 
+    The dicts of left_data and right_data are only
+    used when compatibility-checking uses of the
+    squashing ([!]) operator.
+
     Parameters
     ----------
     left: str
     right: str
+    left_data: dict
+    right_data: dict
     legacy: bool
 
     Returns
     -------
     bool
     """
-    if legacy:
-        if left == right and left[0] not in '> <':
-            return True
-        l, r = left[0], right[0]
-        if (l, r) == ('<', '>') or (l, r) == ('>', '<'):
-            return left[1:] == right[1:]
-        return False
-    else:
-        if left[0] == right[0] == '$' or left[0] == right[0] == '!':
-            return True
 
-        l, r = left[0], right[0]
-        if (l, r) == ('<', '>') or (l, r) == ('>', '<'):
-            return True
+    # Non-legacy only checks the first character for compatibility
+    if not legacy:
+        left = left[0]
+        right = right[0]
+
+    # By flipping the <> for one of the binders, we no longer need special
+    #  case checking downstream
+    left = left[0].translate(str.maketrans('<>','><')) + left[1:]
+
+    if left != right:
         return False
 
+    if left[0] != '!':
+        return True
+
+    # Only the '!' compatibility control left
+    left_name = left_data.get('element', left_data.get('atomname'))
+    right_name = right_data.get('element', right_data.get('atomname'))
+    # Could there be nameless nodes? Anyway, not enough info to reject
```

Comment:
@mnmelo is it safer to check the element or both?

Reply:
I think this is fine as is: atomistic nodes always have an element (unless `*`), and cg nodes will always have an atomname. I think it's a sane idea to assure those match when squashing. Do we need an override ("Do what I tell you, not what you think I want")?